io-sim/test/Test/Control/Monad/Class/MonadMVar.hs

```diff
@@ -0,0 +1,325 @@
+    threadDelay 0.01
+    traverse_ (putMVar v) [1..n]
+    results <- waitAll ts
+    return $ results === [1..n]
+
+prop_takeMVar_fairness_sim :: Positive (Small Int)
+                           -> Property
+prop_takeMVar_fairness_sim (Positive (Small n)) =
+    runSimOrThrow (prop_takeMVar_fairness n)
+
+
+unit_takeMVar_blocks_on_empty
+  :: ( MonadFork  m
+     , MonadDelay m
+     , MonadMVar  m
+     , MonadMonotonicTime m
+     )
+  => m Bool
+unit_takeMVar_blocks_on_empty = do
+    start <- getMonotonicTime
+    let delta = 0.01
+    v :: MVar m () <- newEmptyMVar
+    _ <- forkIO $ threadDelay delta
+               >> putMVar v ()
+    takeMVar v
+    end <- getMonotonicTime
+    return (end `diffTime` start >= delta)
+
+unit_takeMVar_blocks_on_empty_sim :: Assertion
+unit_takeMVar_blocks_on_empty_sim = assertBool "did not block on an empty MVar" $ runSimOrThrow unit_takeMVar_blocks_on_empty
+
+unit_takeMVar_blocks_on_empty_io :: Assertion
+unit_takeMVar_blocks_on_empty_io =
+    unit_takeMVar_blocks_on_empty >>= assertBool "did not block on an empty MVar"
+
+--
+-- tryTakeMVar
+--
+
+
+prop_tryTakeMVar
+  :: ( m ~ IOSim s -- this property is only true in IOSim, since we use time to
+                   -- detect that an operation is non blocking
+     )
+  => Bool -> Bool
+prop_tryTakeMVar isEmpty = runSimOrThrow $ do
+    v <- if isEmpty
+         then newEmptyMVar
+         else newMVar ()
+    start <- getMonotonicTime
+    _ <- tryTakeMVar v
+    end <- getMonotonicTime
+    return $ start == end
+
+unit_tryTakeMVar_empty :: Assertion
+unit_tryTakeMVar_empty = assertBool "blocked on an empty MVar" $
+    prop_tryTakeMVar False
+
+unit_tryTakeMVar_full :: Assertion
+unit_tryTakeMVar_full = assertBool "blocked on an empty MVar" $
+    prop_tryTakeMVar True
+
+
+prop_tryTakeMVar_return_value
+  :: MonadMVar m
+  => Bool
+  -> m Bool
+prop_tryTakeMVar_return_value isEmpty =
+    do v :: MVar m ()
+         <- if isEmpty
+            then newEmptyMVar
+            else newMVar ()
+       a <- tryTakeMVar v
+       return $ isNothing a == isEmpty
+
+unit_tryTakeMVar_return_empty_sim :: Assertion
+unit_tryTakeMVar_return_empty_sim =
+    assertBool "tryTakeMVar on an empty should return result" $
+    runSimOrThrow (prop_tryTakeMVar_return_value True)
+
+unit_tryTakeMVar_return_full_sim :: Assertion
+unit_tryTakeMVar_return_full_sim =
+    assertBool "tryTakeMVar on an full should return result" $
+    runSimOrThrow (prop_tryTakeMVar_return_value False)
+
+--
+-- tryPutMVar
+--
+
+prop_tryPutMVar
+  :: ( m ~ IOSim s -- this property is only true in IOSim, since we use time to
+                   -- detect that an operation is non blocking
+     )
+  => Bool -> Bool
+prop_tryPutMVar isEmpty = runSimOrThrow $ do
+    v <- if isEmpty
+         then newEmptyMVar
+         else newMVar ()
+    start <- getMonotonicTime
+    _ <- tryPutMVar v ()
+    end <- getMonotonicTime
+    return $ start == end
```

Comment:
Another way would be to use `timeout` with a full `MVar` and check that it does not timeout. This way wouldn't be specific to `IOSim`

Reply:
but how then check that `putMVar` doesn't block when the `MVar` is empty?

Reply:
Ah good point, I think the critical case is when the `MVar` is full, if it is empty and we have a test for `tryTakeMVar` not blocking, that covers some part of the behavior we want when the `MVar` is empty. Do we check `take` and `put` are non blocking for the full, empty case, respectively? Then it might be sufficient to make an argument. But this already sounds like we are trying to hard 😛

Reply:
Another way is to check if the trace does not contain any `stm` blocking operation. This would be still `IOSim` sepecific, but a nice example how one could use the low level trace.  But then the test relies on correctness of the `io-sim` tracing mechanism.

Reply:
Fixed.